blackjax/vi/schrodinger_follmer.py

```diff
@@ -0,0 +1,234 @@
+        Log-density of the target distribution
+    step_size
+        Step size of the integration scheme
+    n_samples
+        Number of samples to use to approximate the drift term
+    """
+
+    drift_key, sde_key = jax.random.split(rng_key)
+
+    ravelled_position, unravel_fn = ravel_pytree(state.position)
+    scale = jnp.sqrt(1 - state.time)
+
+    eps_drift = jax.random.normal(drift_key, (n_samples,) + ravelled_position.shape)
+    eps_drift = jax.vmap(unravel_fn)(eps_drift)
+
+    perturbed_position = jax.tree_map(
+        lambda a, b: a[None, ...] + scale * b, state.position, eps_drift
+    )
+
+    log_pdf = jax.vmap(_log_fn_corrected, in_axes=[0, None])(
+        perturbed_position, logdensity_fn
+    )
+    log_pdf -= jnp.max(log_pdf, axis=0, keepdims=True)
+    pdf = jnp.exp(log_pdf)
+
+    num = jax.tree_map(lambda a: pdf @ a, eps_drift)
+    den = scale * jnp.sum(pdf, axis=0)
+
+    drift = jax.tree_map(lambda a: a / den, num)
+
+    eps_sde = jax.random.normal(sde_key, ravelled_position.shape)
+    eps_sde = unravel_fn(eps_sde)
+    next_position = jax.tree_map(
+        lambda a, b, c: a + step_size * b + step_size**0.5 * c,
+        state.position,
+        drift,
+        eps_sde,
+    )
+    next_state = SchrodingerFollmerState(next_position, state.time + step_size)
+    return next_state, SchrodingerFollmerInfo(drift)
+
+
+def sample(
+    rng_key: PRNGKey,
+    initial_state: SchrodingerFollmerState,
+    log_density_fn: Callable,
+    n_steps: int,
+    n_inner_samples,
+    n_samples: int = 1,
+):
+    """
+    Samples from the target distribution using the Schrödinger-Föllmer algorithm.
+
+    Parameters
+    ----------
+    rng_key
+        PRNG key
+    initial_state
+        Current state of the algorithm
+    log_density_fn
+        Log-density of the target distribution
+    n_steps
+        Number of steps to run the algorithm for
+    n_inner_samples
+        Number of samples to use to approximate the drift term
+    n_samples
+        Number of samples to draw
+    """
+    dt = 1.0 / n_steps
+
+    initial_position = initial_state.position
+    if n_samples == 1:
+        initial_states = SchrodingerFollmerState(
+            jax.tree_map(jnp.zeros_like, initial_position), 0.0
+        )
+    else:
+        initial_positions = jax.tree_map(
+            lambda a: jnp.zeros([n_samples, *a.shape]), initial_position
+        )
+        initial_states = SchrodingerFollmerState(
+            initial_positions, jnp.zeros((n_samples,))
+        )
+
+    def body(_, carry):
+        key, states = carry
+        next_key, inner_key = jax.random.split(key)
+        if n_samples == 1:
+            states, _ = step(inner_key, states, log_density_fn, dt, n_inner_samples)
+        else:
+            inner_keys = jax.random.split(inner_key, n_samples)
+            states, _ = jax.vmap(step, [0, 0, None, None, None])(
+                inner_keys, states, log_density_fn, dt, n_inner_samples
+            )
```

Comment:
Good trick, I was struggling with the splitting when n_samples = 1, hence the two cases. Doing it your way should solve it. I modified the unpacking a bit: it was getting a tuple of keys instead of an array.